File: discover-articles.py
#!/usr/bin/env python3
"""
Auto-discovery script for articles (Python version)
Scans the docs/ folder and returns all articles with valid metadata.json
"""

import os
import json
from datetime import datetime
from pathlib import Path

def discover_articles():
    """Scan docs folder and return article metadata"""
    base_folder = Path(__file__).parent / 'docs'
    articles = []
    
    if not base_folder.exists():
        return {
            'error': 'Docs folder not found',
            'articles': []
        }
    
    # Scan all subdirectories in docs/
    for folder in base_folder.iterdir():
        if not folder.is_dir():
            continue
            
        metadata_path = folder / 'metadata.json'
        
        if metadata_path.exists():
            try:
                with open(metadata_path, 'r', encoding='utf-8') as f:
                    metadata = json.load(f)
                
                if 'title' in metadata:
                    metadata['folder'] = folder.name
                    metadata['url'] = f'./docs/{folder.name}/'
                    articles.append(metadata)
                    
            except (json.JSONDecodeError, IOError) as e:
                print(f"Error loading {metadata_path}: {e}")
                continue
    
    # Sort by date (newest first)
    articles.sort(key=lambda x: datetime.fromisoformat(x.get('date', '1970-01-01')), reverse=True)
    
    return {
        'success': True,
        'baseFolder': 'docs',
        'articles': articles,
        'count': len(articles)
    }

if __name__ == '__main__':
    # For CGI or command-line usage
    print('Content-Type: application/json')
    print()
    print(json.dumps(discover_articles(), indent=2))

File: discover-articles.js
#!/usr/bin/env node
/**
 * Auto-discovery script for articles (Node.js version)
 * Scans the docs/ folder and returns all articles with valid metadata.json
 * 
 * Usage:
 * 1. As a standalone script: node discover-articles.js
 * 2. As an Express.js endpoint (see example below)
 */

const fs = require('fs');
const path = require('path');

function discoverArticles() {
    const baseFolder = path.join(__dirname, 'docs');
    const articles = [];
    
    if (!fs.existsSync(baseFolder)) {
        return {
            error: 'Docs folder not found',
            articles: []
        };
    }
    
    try {
        // Read all items in docs folder
        const folders = fs.readdirSync(baseFolder);
        
        folders.forEach(folder => {
            const folderPath = path.join(baseFolder, folder);
            
            // Check if it's a directory
            if (!fs.statSync(folderPath).isDirectory()) {
                return;
            }
            
            // Check for metadata.json
            const metadataPath = path.join(folderPath, 'metadata.json');
            
            if (fs.existsSync(metadataPath)) {
                try {
                    const metadataContent = fs.readFileSync(metadataPath, 'utf8');
                    const metadata = JSON.parse(metadataContent);
                    
                    if (metadata && metadata.title) {
                        metadata.folder = folder;
                        metadata.url = `./docs/${folder}/`;
                        articles.push(metadata);
                    }
                } catch (error) {
                    console.error(`Error loading ${metadataPath}:`, error.message);
                }
            }
        });
        
        // Sort by date (newest first)
        articles.sort((a, b) => new Date(b.date) - new Date(a.date));
        
        return {
            success: true,
            baseFolder: 'docs',
            articles: articles,
            count: articles.length
        };
        
    } catch (error) {
        return {
            error: error.message,
            articles: []
        };
    }
}

// If run directly (not imported as module)
if (require.main === module) {
    const result = discoverArticles();
    console.log(JSON.stringify(result, null, 2));
}

// Export for use in Express.js or other Node apps
module.exports = { discoverArticles };

/* 
 * Example Express.js integration:
 * 
 * const express = require('express');
 * const { discoverArticles } = require('./discover-articles.js');
 * 
 * const app = express();
 * 
 * app.get('/api/articles', (req, res) => {
 *     res.json(discoverArticles());
 * });
 * 
 * app.listen(3000, () => {
 *     console.log('Server running on http://localhost:3000');
 * });
 */
